_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,392 @@
+
+Similarly, postconditions (`#[ensures]`) serve to define the guarantees that a function must uphold after its execution. These ensure the function output or state adheres to the intended expectations, providing an additional safeguard for correctness.
+
+### 2. Generating Verification Harnesses
+
+Sometimes, we need to verify a method for multiple types, which often requires writing numerous harnesses. However, these harnesses usually share the same logic, which makes it inefficient and error-prone to copy and paste similar code repeatedly. This is where [**macros**](https://doc.rust-lang.org/book/ch19-06-macros.html) come in handy. As a further improvement, we can define a single reusable template that dynamically generates harnesses for different types using a single macro, which reduces redundancy and improves maintainability.
+
+In this next code block, we created a macro to generate verification harnesses for each method and type. It showcases the use of macros to generate reusable verification harnesses for different numeric types and methods, ensuring correctness without duplicating code.
+
+```rust
+#[cfg(kani)]
+mod verify {
+    use super::*;
+
+    // Generate harnesses for methods involving two operands
+    // `type`: Integer type
+    // `method`: Method to verify
+    // `harness_name`: A unique identifier for a harness
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>(); // (1)
+                let num2: $type = kani::any::<$type>(); // (1)
+                // (2)
+                unsafe { num1.$method(num2); } // (3)
+                // (4)
+            }
+        }
+    }
+
+    generate_unchecked_math_harness!(i8, unchecked_add, unchecked_add_i8);
+    generate_unchecked_math_harness!(i16, unchecked_add, unchecked_add_i16);
+    // ... Repeat for other integer types (32/64/128-bit/architecture-dependent signed/unsigned integers)
+}
+```
+
+A harness contains several parts:
+
+1. [**Symbolic values**](https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html): `num1` and `num2` are symbolic values generated using `kani::any()`. Kani employs symbolic execution to explore a wide range of input possibilities systematically.
+2. **Assumptions**: With a `#[kani::proof_for_contract]` annotation, Kani automatically inserts `kani::assume()` before the unsafe function call. It ensures that all generated values respect the preconditions of `unchecked_add`. If there are additional assumptions not captured by function contracts, you can specify them manually as well. You can find further details in [Kani&#39;s official RFC for function contracts](https://github.com/model-checking/kani/blob/main/rfc/src/rfcs/0009-function-contracts.md#user-experience).
```

Comment:
```suggestion
2. **Assumptions**: Kani expands `#[requires(precondition)]` to `kani::assume(precondition)` and inserts this assumption in the harness before the unsafe function call. This assumption ensures that all generated values respect the preconditions of `unchecked_add`. If there are additional assumptions not captured by function contracts, you can specify them manually as well. You can find further details in [Kani's blog about function contracts](https://model-checking.github.io/kani-verifier-blog/2024/01/29/function-contracts.html).
```
The presence of the `[kani::proof_for_contract]` does not cause `kani::assume()` to be inserted--if there are no preconditions, there are no assumptions either. The `#[kani::proof_for_contract]` just instructs Kani to check that the postconditions hold given zero or more preconditions.

Reply:
Ah! Thank you for pointing that. Update